### PR TITLE
TileLayer improvements

### PR DESCRIPTION
**Pull request by Pessimistress**
Resolves https://github.com/uber/deck.gl/issues/3511

This PR refactors `BaseTileLayer`/`TileCache` classes to function more similar to `Tile3DLayer`/`Tileset3D`. Traversal now reruns every time any tile finishes loading, instead of waiting for the entire viewport to load. The logic that determines tile visibilities is entirely moved into `TileCache`. It also uses a more sophisticated strategy (configurable by the user) to pick placeholders when a tile is pending.

The new API needs audit.

#### Change List
- Add `strategy` prop to `BaseTileLayer` (see documentation)
- `TileCache` rewrite
- Unit tests

### CODE REVIEW

File: docs/layers/base-tile-layer.md [hunk 1]
```diff
@@ -81,6 +80,15 @@ The maximum cache size for a tile layer. If not defined, it is calculated using
 
 - Default: `null`
 
+##### `strategy` (Enum, optional)
+
+How the tile layer determines the visibility of tiles. One of the following:
+
+* `'default'`: If some tiles in the current viewport are waiting for data to load, aggresively use cached content from other zoom levels to fill the empty space.
+* `'exclusive'`: Avoid showing overlapping tiles when backfilling with cached content.
```
Comment (ibgreen) [inline docs/layers/base-tile-layer.md]: - The two bullets are worded very differently (waiting for data to load/backfilling with cached content) but basically talk about the same thing, this makes it a little confusing.
- I'd prefer if both strategies were named, rather than just calling one `default`.
- in 3d tiles, the strategies are called `ADDITIVE` (all zoom levels drawn at once, used by point clouds) and `REPLACEMENT` (render parent tile until all child tiles are loaded, used by 3d models/photogrammetry).
- The 2d tile semantics are slightly different (i.e. we don't have to draw the entire parent chain in "additive" mode), but if they are close enough then we could call them the same thing, or at least name them in a similar way.
- Basically, I want one mode where we draw a parent tile and wait for all 4 children to load - before that we draw the children (and stop drawing the parent).

Since "default/ADDITIVE" doesn't work opacity is not 1, I propose that the mode would automatically be set to `REPLACEMENT` in those cases.

File: docs/layers/base-tile-layer.md [hunk 2]
```diff
@@ -81,6 +80,15 @@ The maximum cache size for a tile layer. If not defined, it is calculated using
...
+
+- Default: `BaseTileLayer.STRATEGY_DEFAULT`
```
Comment (ibgreen) [inline docs/layers/base-tile-layer.md]: The bulleted list has strings. This uses an enum that is a static member of the class, which seems unusual for our API and may require extra wiring to use declaratively.

File: modules/layers/src/base-tile-layer/base-tile-layer.js [hunk 3]
```diff
@@ -7,14 +7,15 @@ const defaultProps = {
   renderSubLayers: {type: 'function', value: props => new GeoJsonLayer(props), compare: false},
   getTileData: {type: 'function', value: ({x, y, z}) => Promise.resolve(null), compare: false},
   tileToBoundingBox: {type: 'function', value: () => null, compare: false},
-  getTileIndices: {type: 'function', value: () => null, compare: false},
+  getTileIndices: {type: 'function', value: () => [], compare: false},
```
Comment (ibgreen) [inline modules/layers/src/base-tile-layer/base-tile-layer.js]: Nit; For self-documentation purposes, can we show here what parameters this function takes?

File: modules/layers/src/base-tile-layer/utils/tile-cache.js [hunk 4]
```diff
@@ -1,5 +1,33 @@
...
+   show cached parent tile if children are loading
+   +-----------+       +-----+            +-----+-----+
+   |           |       |     |            |     |     |
+   |           |       |     |            |     |     |
+   |           |  -->  +-----+-----+  ->  +-----+-----+
+   |           |             |     |      |     |     |
+   |           |             |     |      |     |     |
+   +-----------+             +-----+      +-----+-----+
+
+   show cached children tiles when parent is loading
```
Comment (ibgreen) [inline modules/layers/src/base-tile-layer/utils/tile-cache.js]: Two way reuse, both up and down... This is not used by 3d tiles. Does this complicate things much?

File: modules/layers/src/base-tile-layer/utils/tile-cache.js [hunk 5]
```diff
@@ -1,5 +1,33 @@
...
+ */
+const TILE_STATE_PLACEHOLDER = 3;
+const TILE_STATE_HIDDEN = 4;
+// tiles that should be displayed in the current viewport
+const TILE_STATE_SELECTED = 5;
+
+export const STRATEGY_EXCLUSIVE = 'exclusive';
+export const STRATEGY_DEFAULT = 'default';
+
 /**
```
Comment (ibgreen) [inline modules/layers/src/base-tile-layer/utils/tile-cache.js]: This file looks quite complicated, and happens to be a piece where Unfolded needs to do a lot of work. It would be good to see if we can improve the structure a bit. 

As my initial reaction, I would prefer aligning this code with the class structure in the 3D tiles implementation, namely:

- A `Tileset2D` class that accepts a viewport and return a list of loaded tiles to display, and also accepts callbacks to call when new tiles load. 
- Thus the TileLayer would take a `Tileset2D` instance and we could move most differences between various tile loading strategies out of the `Tileset2D` layer.
- A separate `Tile2DCache` class that handles any actual caching.

I am not yet 100% sure if I am mainly just proposing renaming this class, or if there is meaningful simpliciation/decomposition that can be done this way.

File: modules/layers/src/base-tile-layer/utils/tile.js [hunk 6]
```diff
@@ -5,10 +5,13 @@ export default class Tile {
     this.z = z;
     this.bbox = tileToBoundingBox(this.x, this.y, this.z);
     this.isVisible = true;
-    this.getTileData = getTileData;
```
Comment (ibgreen) [inline modules/layers/src/base-tile-layer/utils/tile.js]: This class is called `Tile3DHeader` in 3d tiles, as it is not the tile contents but the metadata. Could it make sense to align?

Comment (Pessimistress) [reply modules/layers/src/base-tile-layer/utils/tile.js]: This class contains both the metadata and the content. Where is the content stored in 3d tiles?

File: modules/layers/src/base-tile-layer/utils/tile-cache.js (hunk 4 above)
Comment (Pessimistress) [reply modules/layers/src/base-tile-layer/utils/tile-cache.js]: 3d-tiles loads everything from the root to the current level, so parent is always available. 2d only loads tiles at the current level. When you zoom out and no ancestors are available, the map will flash white and you lose all visual reference. Yes it adds complexity, but it's also needed for usability.

File: docs/layers/base-tile-layer.md (hunk 1 above)
Comment (Pessimistress) [reply docs/layers/base-tile-layer.md]: I'm reluctant to use the name `additive` because it's not the same thing as what 3d tiles uses.

How do you feel about calling this prop `replacementStrategy`, with two possible values:

- `'coverage'`: prioritize showing the best available content for every pixel
- `'no-overlap'`: avoid showing overlapping tiles

**Comment by Pessimistress** [comment]
Regarding the `BaseTileLayer` class: we need an API audit before 8.1 release for sure. If we are not moving it into the core layers, then there will be only one `TileLayer` as before, just exposing a few new props to enable the new use cases.

I'm ok with this arrangement for now, given this non-geospatial support is new. In the long term, do we want to investigate non-geospatial support for 3d tiles (e.g. Potree is technically not bound to geospatial)? If so, we may want to move these two layers to a new module `@deck.gl/tile-layers` and drop any specific format dependency?
